File: 2022/day25.cpp
// https://adventofcode.com/2022/day/25

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <cmath>
#include <print>
#include <ranges>

fn parse(String const& filename) -> Vec<String> {
  return aoc::read_lines(filename);
}

fn snafu_char_to_int(char value) -> int {
  switch (value) {
    case '-':
      return -1;
    case '=':
      return -2;
    default:
      AOC_ASSERT((value >= '0') && (value <= '2'),
                 "Invalid character provided");
      return static_cast<int>(value - '0');
  }
}

fn int_to_snafu_char(int value) -> char {
  switch (value) {
    case -1:
      return '-';
    case -2:
      return '=';
    default:
      AOC_ASSERT((value >= 0) && (value <= 2), "Invalid integer provided");
      return static_cast<char>(value + '0');
  }
}

template <int base>
fn to_decimal(str snafu) -> i64 {
  i64 total = 0;
  for (i64 multiplier = 1;
       i64 digit :
       snafu | stdv::reverse | stdv::transform(&snafu_char_to_int)) {
    total += digit * multiplier;
    multiplier *= base;
  }
  return total;
}

template <int base>
fn sum_snafus(str lhs, str rhs) -> String {
  auto output = String{};
  output.resize(lhs.size() + rhs.size() + 1);
  stdr::fill(output, '0');
  let lhs_reversed = lhs | stdv::reverse;
  let rhs_reversed = rhs | stdv::reverse;
  auto output_reversed = output | stdv::reverse;
  let min_len = static_cast<int>(std::min(lhs.size(), rhs.size()));
  let max_len = static_cast<int>(std::max(lhs.size(), rhs.size()));
  int carry = 0;
  int idx = 0;
  for (; idx < min_len; ++idx) {
    int digit_sum = snafu_char_to_int(lhs_reversed[idx]) +
                    snafu_char_to_int(rhs_reversed[idx]) +
                    carry;
    carry =
        static_cast<int>(std::lround(static_cast<double>(digit_sum) / base));
    output_reversed[idx] = int_to_snafu_char(digit_sum - (carry * base));
  }
  let longer_reversed =
      (lhs.size() >= rhs.size()) ? lhs_reversed : rhs_reversed;
  for (; idx < max_len; ++idx) {
    int digit_sum = snafu_char_to_int(longer_reversed[idx]) + carry;
    carry =
        static_cast<int>(std::lround(static_cast<double>(digit_sum) / base));
    output_reversed[idx] = int_to_snafu_char(digit_sum - (carry * base));
  }
  output_reversed[idx] = int_to_snafu_char(carry);
  output = String{aoc::ltrim(output, "0")};
  return output;
}

template <int base>
fn to_snafu(const i64 number) -> String {
  int size = aoc::num_digits(number);
  i64 multiplier = aoc::pown<i64>(base, size - 1);
  do {
    ++size;
    multiplier *= base;
  } while ((number / multiplier) > 0);

  auto snafu = String{};
  snafu.resize(size);

  i64 remaining = number;
  for (auto& output_char : snafu) {
    let digit = static_cast<int>(
        std::lround(static_cast<double>(remaining) / multiplier));
    output_char = int_to_snafu_char(digit);
    remaining -= digit * multiplier;
    multiplier /= base;
  }
  return String{aoc::ltrim(snafu, "0")};
}

fn solve(Vec<String> const& input) -> String {
  return stdr::fold_left(input | stdv::drop(1), input[0], sum_snafus<5>);
}

#define AOC_EXPECT_SAME_CONVERSION(base, decimal, snafu)                       \
  {                                                                            \
    AOC_EXPECT_RESULT(snafu, to_snafu<base>(decimal));                         \
    static_assert(decimal == to_decimal<base>(snafu));                         \
  }

int main() {
  // Unit tests

  AOC_EXPECT_RESULT("1=21", (sum_snafus<5>(to_snafu<5>(49), to_snafu<5>(37))));

  AOC_EXPECT_SAME_CONVERSION(5, 976, "2=-01");
  AOC_EXPECT_SAME_CONVERSION(5, 1747, "1=-0-2");
  AOC_EXPECT_SAME_CONVERSION(5, 906, "12111");
  AOC_EXPECT_SAME_CONVERSION(5, 198, "2=0=");
  AOC_EXPECT_SAME_CONVERSION(5, 11, "21");
  AOC_EXPECT_SAME_CONVERSION(5, 201, "2=01");
  AOC_EXPECT_SAME_CONVERSION(5, 31, "111");
  AOC_EXPECT_SAME_CONVERSION(5, 1257, "20012");
  AOC_EXPECT_SAME_CONVERSION(5, 32, "112");
  AOC_EXPECT_SAME_CONVERSION(5, 353, "1=-1=");
  AOC_EXPECT_SAME_CONVERSION(5, 107, "1-12");
  AOC_EXPECT_SAME_CONVERSION(5, 7, "12");
  AOC_EXPECT_SAME_CONVERSION(5, 3, "1=");
  AOC_EXPECT_SAME_CONVERSION(5, 37, "122");

  std::println("Part 1");
  let example = parse("day25.example");
  AOC_EXPECT_RESULT("2=-1=0", solve(example));
  let input = parse("day25.input");
  AOC_EXPECT_RESULT("2-=12=2-2-2-=0012==2", solve(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day25.rs
fn parse(filename: &str) -> Vec<String> {
    aoc::file::read_lines(filename)
}

fn snafu_char_to_int(ch: char) -> i64 {
    match ch {
        '-' => -1,
        '=' => -2,
        '0'..='2' => (ch as u8 - b'0') as i64,
        _ => panic!("Invalid SNAFU character: {ch}"),
    }
}

fn int_to_snafu_char(value: i64) -> char {
    match value {
        -1 => '-',
        -2 => '=',
        0..=2 => (b'0' + value as u8) as char,
        _ => panic!("Invalid SNAFU digit: {value}"),
    }
}

fn to_decimal(snafu: &str) -> i64 {
    snafu
        .chars()
        .rev()
        .zip(std::iter::successors(Some(1i64), |&m| Some(m * 5)))
        .map(|(ch, multiplier)| snafu_char_to_int(ch) * multiplier)
        .sum()
}

fn sum_snafus(lhs: &str, rhs: &str) -> String {
    let lhs_reversed: Vec<char> = lhs.chars().rev().collect();
    let rhs_reversed: Vec<char> = rhs.chars().rev().collect();
    let max_len = lhs_reversed.len().max(rhs_reversed.len());
    let mut carry = 0i64;
    let mut output = Vec::with_capacity(max_len + 1);
    for idx in 0..=max_len {
        let lhs_digit = lhs_reversed
            .get(idx)
            .map(|&ch| snafu_char_to_int(ch))
            .unwrap_or(0);
        let rhs_digit = rhs_reversed
            .get(idx)
            .map(|&ch| snafu_char_to_int(ch))
            .unwrap_or(0);
        let digit_sum = lhs_digit + rhs_digit + carry;
        carry = (digit_sum as f64 / 5.0).round() as i64;
        output.push(int_to_snafu_char(digit_sum - carry * 5));
    }
    output.reverse();
    output
        .into_iter()
        .collect::<String>()
        .trim_start_matches('0')
        .to_string()
}

fn to_snafu(mut number: i64) -> String {
    let mut digits = Vec::new();
    while number != 0 {
        let remainder = number.rem_euclid(5);
        let (digit, carry) = match remainder {
            3 => (-2i64, 1i64),
            4 => (-1, 1),
            r => (r, 0),
        };
        digits.push(int_to_snafu_char(digit));
        number = number / 5 + carry;
    }
    if digits.is_empty() {
        "0".to_string()
    } else {
        digits.into_iter().rev().collect()
    }
}

fn solve(input: &[String]) -> String {
    input
        .iter()
        .skip(1)
        .fold(input[0].clone(), |acc, rhs| sum_snafus(&acc, rhs))
}

fn main() {
    // Unit tests
    aoc::expect_result!("1=21", sum_snafus(&to_snafu(49), &to_snafu(37)));

    aoc::expect_result!("2=-01", to_snafu(976));
    aoc::expect_result!(976_i64, to_decimal("2=-01"));
    aoc::expect_result!("1=-0-2", to_snafu(1747));
    aoc::expect_result!(1747_i64, to_decimal("1=-0-2"));
    aoc::expect_result!("12111", to_snafu(906));
    aoc::expect_result!(906_i64, to_decimal("12111"));
    aoc::expect_result!("2=0=", to_snafu(198));
    aoc::expect_result!(198_i64, to_decimal("2=0="));
    aoc::expect_result!("21", to_snafu(11));
    aoc::expect_result!(11_i64, to_decimal("21"));
    aoc::expect_result!("2=01", to_snafu(201));
    aoc::expect_result!(201_i64, to_decimal("2=01"));
    aoc::expect_result!("111", to_snafu(31));
    aoc::expect_result!(31_i64, to_decimal("111"));
    aoc::expect_result!("20012", to_snafu(1257));
    aoc::expect_result!(1257_i64, to_decimal("20012"));
    aoc::expect_result!("112", to_snafu(32));
    aoc::expect_result!(32_i64, to_decimal("112"));
    aoc::expect_result!("1=-1=", to_snafu(353));
    aoc::expect_result!(353_i64, to_decimal("1=-1="));
    aoc::expect_result!("1-12", to_snafu(107));
    aoc::expect_result!(107_i64, to_decimal("1-12"));
    aoc::expect_result!("12", to_snafu(7));
    aoc::expect_result!(7_i64, to_decimal("12"));
    aoc::expect_result!("1=", to_snafu(3));
    aoc::expect_result!(3_i64, to_decimal("1="));
    aoc::expect_result!("122", to_snafu(37));
    aoc::expect_result!(37_i64, to_decimal("122"));

    println!("Part 1");
    let example = parse("day25.example");
    aoc::expect_result!("2=-1=0", solve(&example));
    let input = parse("day25.input");
    aoc::expect_result!("2-=12=2-2-2-=0012==2", solve(&input));
}
